src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -546,6 +547,11 @@ public void OnTrailer(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
     private void IncrementRequestHeadersCount()
     {
         _requestHeadersParsed++;
+        CheckRequestHeadersCountLimit();
+    }
+
+    protected virtual void CheckRequestHeadersCountLimit(bool final = false)
+    {
         if (_requestHeadersParsed > ServerOptions.Limits.MaxRequestHeaderCount)
```

Comment:
Let's add a new field for `_incrementalMaxRequestHeaderCount` or `_softMaxRequestHeaderCount` and compare against that in the existing non-virtual `IncrementRequestHeadersCount()` method. The derived classes can adjust this as appropriate. This is a really hot path.